### ISSUE TITLE
NullPointerException when calling Transformer.transform()

### ISSUE THREAD

**Issue by gsiewe** [combust/mleap#328]
Hi,

we use mleap to run a Spark MPC Model. Occasionally we get a nullpointer exception at classification runtime. The nullpointer is not reproducable when we send the exact same input data again. Could it be that the model is not threadsafe and we run into a racecondition? Every hint is much appreciated, we are trying to debug this now for quite some time.

Stacktrace is really long and attached as file [stacktrace.txt](https://github.com/combust/mleap/files/1660727/stacktrace.txt)

And this is our code to run the model:

    class MPCClassifier @Inject()(schemaProvider: SchemaProvider,
                                  @Named("mpc-model") modelLocation: ModelLocation,
                                  metricRegistry: MetricRegistry)(implicit ec: ExecutionContext)
        extends Classifier[MpcResult] {

      override val model: Transformer = {
        (for (bundleFile ← managed(BundleFile(Paths.get(modelLocation.location).toUri))) yield {
          bundleFile.loadMleapBundle().get.root
        }).opt.get
      }

      override def classify(row: Row): Future[MpcResult] = Future {
        val leapFrame = buildLeapFrame(row, schemaProvider.schema)

        val inputPrediction: Try[ClassificationResult] = for {
          dataset <- model.transform(leapFrame)
          predictionColumn <- dataset.select("prediction")
          predictionRow <- Try(predictionColumn.dataset.head)
          prediction <- Try(predictionRow.toSeq.head)
        } yield ClassificationResult.forPrediction(prediction.asInstanceOf[Double])
      }
    }

**Comment by MeiSign** [comment]
The above problem happens with mleap 0.9.0. We have another mpc classifier running on version 0.7.0 and it is not showing this behaviour although we use the same implementation of the model usage.

**Comment by polya20** [comment]
This is true, I had to downgrade from 0.8 to 0.7 to get things working.

**Comment by hollinwilkins** [comment]
Hey everyone,

This seems like a big issue. Would definitely like you all to be able to move up past MLeap 0.7.0. Has anyone discovered where the error could be coming from in your investigations?

**Comment by hollinwilkins** [comment]
@gsiewe @polya20 @MeiSign Can you verify this is still happening? Would like to plan a fix for this.

**Comment by MeiSign** [comment]
Happening with the new version you mean? We still have 0.9.0 in use and it is still happening. We worked around with retry the classification 3 times if it fails.

**Comment by gsiewe** [comment]
We upgraded to version 0.10.0 and the issue is still happening.
Any updates regarding this issue?

**Comment by gsiewe** [comment]
We upgraded to version 0.10.0 and the issue is still happening (check the stack trace
[stacktracet_version_0.10.0.txt](https://github.com/combust/mleap/files/2274708/stacktracet_version_0.10.0.txt)). 

We had a closer look at what is happening when we call the method ml.combust.mleap.runtime.frame.Transformer#transform and we could see that the class ml.combust.mleap.core.ann.FeedForwardModel and specially the methods  ml.combust.mleap.core.ann.FeedForwardModel#forward and ml.combust.mleap.core.ann.FeedForwardModel#computeGradient are not thread safe. In the case multiple threads execute the method ml.combust.mleap.core.ann.FeedForwardModel#forward we can have concurrency problems because the variable **outputs** is updated without any thread synchronization. And this is exactly the case for us, the method ml.combust.mleap.core.ann.FeedForwardModel#forward can be executed by multiple threads concurrently.

We are loading the model only once when the service starts, we therefore have one instance of the class ml.combust.mleap.runtime.transformer.classification.MultiLayerPerceptronClassifier with a reference to an instance of the class ml.combust.mleap.core.classification.MultiLayerPerceptronClassifierModel which in turn is holding an instance of the class ml.combust.mleap.core.ann.FeedForwardModel. Calling the method ml.combust.mleap.core.ann.FeedForwardModel#forward concurrently as explained above causes sometimes a NullPointerException in the evaluation of the first **IF-Statement** in the method, specially in the evaluation of the second part of the first **IF-Statement** in the method.

We are loading the model as follow:

`lazy val model: Transformer = {
    val bundle =
      (for (bundleFile ← managed(BundleFile(Paths.get("mpc-model").toUri)))
        yield {
          bundleFile.loadMleapBundle().get.root
        }).opt.get
    bundle
  }`

We would like to suggest following code changes to solve the problem:
The variable **outputs** should become a local variable of the method ml.combust.mleap.core.ann.FeedForwardModel#forward.

`override def forward(data: BDM[Double], includeLastLayer: Boolean): Array[BDM[Double]] = {
    var outputs: Array[BDM[Double]] = null
    // Initialize output arrays for all layers. Special treatment for InPlace
    val currentBatchSize = data.cols
    // TODO: allocate outputs as one big array and then create BDMs from it
    if (outputs == null || outputs(0).cols != currentBatchSize) {
      outputs = new Array[BDM[Double]](layers.length)
      var inputSize = data.rows
      for (i <- 0 until layers.length) {
        if (layers(i).inPlace) {
          outputs(i) = outputs(i - 1)
        } else {
          val outputSize = layers(i).getOutputSize(inputSize)
          outputs(i) = new BDM[Double](outputSize, currentBatchSize)
          inputSize = outputSize
        }
      }
    }
    layerModels(0).eval(data, outputs(0))
    val end = if (includeLastLayer) layerModels.length else layerModels.length - 1
    for (i <- 1 until end) {
      layerModels(i).eval(outputs(i - 1), outputs(i))
    }
    outputs
  }`

The variable **deltas** should become a local variable of the method 
ml.combust.mleap.core.ann.FeedForwardModel#computeGradient.

`override def computeGradient(
                                data: BDM[Double],
                                target: BDM[Double],
                                cumGradient: Vector,
                                realBatchSize: Int): Double = {
    var deltas: Array[BDM[Double]] = null
    val outputs = forward(data, true)
    val currentBatchSize = data.cols
    // TODO: allocate deltas as one big array and then create BDMs from it
    if (deltas == null || deltas(0).cols != currentBatchSize) {
      deltas = new Array[BDM[Double]](layerModels.length)
      var inputSize = data.rows
      for (i <- 0 until layerModels.length - 1) {
        val outputSize = layers(i).getOutputSize(inputSize)
        deltas(i) = new BDM[Double](outputSize, currentBatchSize)
        inputSize = outputSize
      }
    }
    val L = layerModels.length - 1
    // TODO: explain why delta of top layer is null (because it might contain loss+layer)
    val loss = layerModels.last match {
      case levelWithError: LossFunction => levelWithError.loss(outputs.last, target, deltas(L - 1))
      case _ =>
        throw new UnsupportedOperationException("Top layer is required to have objective.")
    }
    for (i <- (L - 2) to (0, -1)) {
      layerModels(i + 1).computePrevDelta(deltas(i + 1), outputs(i + 1), deltas(i))
    }
    val cumGradientArray = cumGradient.toArray
    var offset = 0
    for (i <- 0 until layerModels.length) {
      val input = if (i == 0) data else outputs(i - 1)
      layerModels(i).grad(deltas(i), input,
        new BDV[Double](cumGradientArray, offset, 1, layers(i).weightSize))
      offset += layers(i).weightSize
    }
    loss
  }`

With this in place the class ml.combust.mleap.core.ann.FeedForwardModel becomes thread safe.

**Comment by hollinwilkins** [comment]
@gsiewe Thank you for digging into this issue, great find. Can you submit a PR with the changes?

**Comment by gsiewe** [comment]
ok i will submit a PR.

**Comment by gsiewe** [comment]
Hi @hollinwilkins,

i submit the pull request. Can you may be check it.

**Comment by ancasarb** [comment]
PR has been merged, closing this issue.
